File: lwk_bindings/tests/bindings/authenticated_esplora_client.py
#!/usr/bin/env python3
"""Authenticated Esplora client with LWK (OAuth2 / static token / headers).

Some Esplora servers — e.g. Blockstream Enterprise — require authentication.
The bindings expose this through `EsploraClientBuilder`'s `token_provider` and
`headers` fields.

This script does two things:

  * Offline checks (always run): build authenticated clients and assert the
    builder wiring behaves as expected. No network or credentials needed,
    because the OAuth token is only fetched on the first request.
  * Live check (opt-in): if CLIENT_ID and CLIENT_SECRET are set, actually hit
    an authenticated endpoint and assert we can read the chain tip. Point it at
    a different host/login with ESPLORA_BASE_URL / ESPLORA_LOGIN_URL.
"""

import logging
import os
import sys

from lwk import *

logging.basicConfig(level=logging.INFO, format="%(levelname)s %(message)s")
log = logging.getLogger("authenticated_esplora")

network = Network.mainnet()

client_id = os.environ.get("CLIENT_ID", "your_client_id")
client_secret = os.environ.get("CLIENT_SECRET", "your_client_secret")


def offline_checks():
    """Build authenticated clients and assert the wiring, without any network."""
    # ANCHOR: authenticated_esplora_client
    base_url = "https://enterprise.blockstream.info/liquid/api"
    login_url = "https://login.blockstream.com/realms/blockstream-public/protocol/openid-connect/token"

    builder = EsploraClientBuilder(
        base_url=base_url,
        network=network,
        token_provider=TokenProvider.BLOCKSTREAM(
            url=login_url,
            client_id=client_id,
            client_secret=client_secret,
        ),
    )
    client = EsploraClient.from_builder(builder)
    # ANCHOR_END: authenticated_esplora_client

    # The OAuth2 provider is recorded on the builder and the client is built.
    assert builder.token_provider.is_BLOCKSTREAM()
    assert client is not None
    log.info("built OAuth2 (Blockstream) client for %s", base_url)

    # A static token (when you already have one) and/or custom headers also work.
    builder = EsploraClientBuilder(
        base_url=base_url,
        network=network,
        token_provider=TokenProvider.STATIC(token="my-token"),
        headers={"X-Custom-Header": "value"},
    )
    client = EsploraClient.from_builder(builder)

    assert builder.token_provider.is_STATIC()
    assert builder.token_provider.token == "my-token"
    assert builder.headers == {"X-Custom-Header": "value"}
    assert client is not None
    log.info("built static-token client with custom headers")

    # No-auth builders must keep working (auth fields default to None).
    builder = EsploraClientBuilder(base_url=base_url, network=network)
    assert builder.token_provider is None
    assert builder.headers is None
    assert EsploraClient.from_builder(builder) is not None
    log.info("offline checks passed")


def live_check():
    """If credentials are provided, prove auth works by reading the chain tip."""
    if client_id == "your_client_id" or client_secret == "your_client_secret":
        log.info("skipping live check: set CLIENT_ID and CLIENT_SECRET to enable it")
        return

    base_url = os.environ.get(
        "ESPLORA_BASE_URL", "https://enterprise.uat.blockstream.info/liquid/api"
    )
    login_url = os.environ.get(
        "ESPLORA_LOGIN_URL",
        "https://login.uat.blockstream.com/realms/blockstream-public/protocol/openid-connect/token",
    )

    builder = EsploraClientBuilder(
        base_url=base_url,
        network=network,
        token_provider=TokenProvider.BLOCKSTREAM(
            url=login_url, client_id=client_id, client_secret=client_secret
        ),
        timeout=30,
    )
    client = EsploraClient.from_builder(builder)

    log.info("authenticating against %s ...", base_url)
    tip = client.tip()  # triggers the OAuth token fetch + an authenticated request
    log.info("authenticated OK: chain tip height=%d hash=%s", tip.height(), tip.block_hash())
    assert tip.height() > 100, f"unexpected tip height {tip.height()}"


def main():
    offline_checks()
    try:
        live_check()
    except LwkError as e:
        # A network/credential failure here shouldn't look like a wiring bug.
        log.error("live check failed (network or credentials): %s", e)
        return 1
    log.info("done")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: lwk_bindings/go/authenticated_esplora_client.go
package main

import (
	"fmt"
	"log"
	"os"

	"lwk-go/lwk"
)

func offlineChecks(clientId, clientSecret string, network *lwk.Network) {
	// ANCHOR: authenticated_esplora_client
	baseUrl := "https://enterprise.blockstream.info/liquid/api"
	loginUrl := "https://login.blockstream.com/realms/blockstream-public/protocol/openid-connect/token"

	var tpBlockstream lwk.TokenProvider = lwk.TokenProviderBlockstream{
		Url:          loginUrl,
		ClientId:     clientId,
		ClientSecret: clientSecret,
	}

	builder := lwk.EsploraClientBuilder{
		BaseUrl:       baseUrl,
		Network:       network,
		TokenProvider: &tpBlockstream,
	}
	client, err := lwk.EsploraClientFromBuilder(builder)
	// ANCHOR_END: authenticated_esplora_client
	if err != nil {
		log.Fatal(err)
	}

	// The OAuth2 provider is recorded on the builder and the client is built.
	if _, ok := (*builder.TokenProvider).(lwk.TokenProviderBlockstream); !ok {
		panic("expected builder.TokenProvider to be of type TokenProviderBlockstream")
	}
	if client == nil {
		panic("client is nil")
	}
	log.Printf("built OAuth2 (Blockstream) client for %s", baseUrl)

	// A static token (when you already have one) and/or custom headers also work.
	var tpStatic lwk.TokenProvider = lwk.TokenProviderStatic{Token: "my-token"}
	headers := map[string]string{"X-Custom-Header": "value"}
	builderStatic := lwk.EsploraClientBuilder{
		BaseUrl:       baseUrl,
		Network:       network,
		TokenProvider: &tpStatic,
		Headers:       &headers,
	}
	clientStatic, err := lwk.EsploraClientFromBuilder(builderStatic)
	if err != nil {
		log.Fatal(err)
	}

	staticProvider, ok := (*builderStatic.TokenProvider).(lwk.TokenProviderStatic)
	if !ok {
		panic("expected builder.TokenProvider to be of type TokenProviderStatic")
	}
	if staticProvider.Token != "my-token" {
		panic("expected token to be 'my-token'")
	}
	if builderStatic.Headers == nil || (*builderStatic.Headers)["X-Custom-Header"] != "value" {
		panic("expected custom headers to be set")
	}
	if clientStatic == nil {
		panic("client is nil")
	}
	log.Println("built static-token client with custom headers")

	// No-auth builders must keep working (auth fields default to nil).
	builderNoAuth := lwk.EsploraClientBuilder{
		BaseUrl: baseUrl,
		Network: network,
	}

	if builderNoAuth.TokenProvider != nil {
		panic("expected TokenProvider to be nil")
	}
	if builderNoAuth.Headers != nil {
		panic("expected Headers to be nil")
	}

	clientNoAuth, err := lwk.EsploraClientFromBuilder(builderNoAuth)
	if err != nil {
		log.Fatal(err)
	}
	if clientNoAuth == nil {
		panic("client is nil")
	}
	log.Println("offline checks passed")
}

func liveCheck(clientId, clientSecret string, network *lwk.Network) error {
	if clientId == "your_client_id" || clientSecret == "your_client_secret" {
		log.Println("skipping live check: set CLIENT_ID and CLIENT_SECRET to enable it")
		return nil
	}

	baseUrl := os.Getenv("ESPLORA_BASE_URL")
	if baseUrl == "" {
		baseUrl = "https://enterprise.uat.blockstream.info/liquid/api"
	}

	loginUrl := os.Getenv("ESPLORA_LOGIN_URL")
	if loginUrl == "" {
		loginUrl = "https://login.uat.blockstream.com/realms/blockstream-public/protocol/openid-connect/token"
	}

	timeout := uint8(30)

	var liveTp lwk.TokenProvider = lwk.TokenProviderBlockstream{
		Url:          loginUrl,
		ClientId:     clientId,
		ClientSecret: clientSecret,
	}

	builder := lwk.EsploraClientBuilder{
		BaseUrl:       baseUrl,
		Network:       network,
		TokenProvider: &liveTp,
		Timeout:       &timeout,
	}

	client, err := lwk.EsploraClientFromBuilder(builder)
	if err != nil {
		return err
	}

	log.Printf("authenticating against %s ...", baseUrl)

	// triggers the OAuth token fetch + an authenticated request
	tip, err := client.Tip()
	if err != nil {
		return err
	}

	log.Printf("authenticated OK: chain tip height=%d hash=%s", tip.Height(), tip.BlockHash())
	if tip.Height() <= 100 {
		return fmt.Errorf("unexpected tip height %d", tip.Height())
	}

	return nil
}

func main() {
	network := lwk.NetworkMainnet()

	clientId := os.Getenv("CLIENT_ID")
	if clientId == "" {
		clientId = "your_client_id"
	}

	clientSecret := os.Getenv("CLIENT_SECRET")
	if clientSecret == "" {
		clientSecret = "your_client_secret"
	}

	offlineChecks(clientId, clientSecret, network)

	if err := liveCheck(clientId, clientSecret, network); err != nil {
		// A network/credential failure here shouldn't look like a wiring bug.
		log.Printf("live check failed (network or credentials): %v\n", err)
		os.Exit(1)
	}

	log.Println("done")
}
